### PR TITLE
[Distributed] Add roles

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
We need to distinguish roles of a client process. The PR moves roles from kvstore and create a component to keep track of the roles of client processes.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/distributed/role.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            kv_store.barrier_count[self.role] = 0
+        role[self.role].add((self.client_id, self.machine_id))
+        total_count = 0
+        for key in role:
+            total_count += len(role[key])
+        # Clients are blocked util all clients register their roles.
+        if total_count == kv_store.num_clients:
+            res_list = []
+            for target_id in range(kv_store.num_clients):
+                print('send', target_id)
```
Comment (aksnzhy) [inline python/dgl/distributed/role.py:62]: Remove this print?

File: python/dgl/distributed/role.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    def __getstate__(self):
+        return self.client_id, self.machine_id, self.role
+
+    def __setstate__(self, state):
+        self.client_id, self.machine_id, self.role = state
+
+    def process_request(self, server_state):
+        kv_store = server_state.kv_store
+        role = kv_store.role
+        if self.role not in role:
```
Comment (aksnzhy) [inline python/dgl/distributed/role.py:51]: It's better to put role in rpc space, instead of kvstore.

File: python/dgl/distributed/role.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    Each process is associated with a role so that we can determine what
+    function can be invoked in a process. For example, we do not allow some
+    functions in sampler processes.
+
+    The initialization includes registeration the role of the current process and
+    get the roles of all client processes. It also computes the rank of all client
+    processes in a deterministic way so that all clients will have the same rank for
+    the same client process.
+    """
+    global CUR_ROLE
```
Comment (aksnzhy) [inline python/dgl/distributed/role.py:119]: Consiter put `CUR_ROLE` into `RPCContext` in` rpc.h `

File: python/dgl/distributed/role.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    response = rpc.recv_response()
+    assert response.msg == GET_ROLE_MSG
+
+    # Here we want to compute a new rank for each client.
+    # We compute the per-role rank as well as global rank.
+    # For per-role rank, we ensure that all ranks within a machine is contiguous.
+    # For global rank, we also ensure that all ranks within a machine are contiguous,
+    # and all ranks within a role are contiguous.
+    global PER_ROLE_RANK
+    global GLOBAL_RANK
```
Comment (aksnzhy) [inline python/dgl/distributed/role.py:141]: Consiter put these two into RPCContext.

File: python/dgl/distributed/role.py (hunk 3 above)
Comment (zheng-da) [reply python/dgl/distributed/role.py:119]: why?

File: python/dgl/distributed/role.py (hunk 4 above)
Comment (zheng-da) [reply python/dgl/distributed/role.py:141]: these are fairly complex data structures. i'm not sure if role affects the behavior of RPC.

**Pull request merged**